src/common/nvlist_utils.c

```diff
@@ -436,3 +436,378 @@ PyObject *py_dump_nvlist(nvlist_t *nvl, boolean_t json)
+static
+int add_leaf_vdev(nvlist_t *item, PyObject *path)
+{
+	char rpath[PATH_MAX];
+	struct stat64 statbuf;
+	const char *cpath = PyUnicode_AsUTF8(path);
+	if (!cpath) {
+		PyErr_SetString(PyExc_RuntimeError,
+			"Failed to create C string from Python string");
+		return (-1);
+	}
+	if (realpath(cpath, rpath) == NULL) {
+		PyErr_Format(PyExc_RuntimeError,
+			"Cannot resolve path %s", cpath);
+		return (-1);
+	}
+	if (stat64(rpath, &statbuf) != 0) {
+		PyErr_Format(PyExc_RuntimeError,
+			"Cannot open %s", rpath);
+		return (-1);
+	}
+	if (S_ISBLK(statbuf.st_mode)) {
+		fnvlist_add_string(item, "type", "disk");
+	} else if (S_ISREG(statbuf.st_mode)) {
+		fnvlist_add_string(item, "type", "file");
+	} else {
+		PyErr_Format(PyExc_RuntimeError,
+		    "%s is not a block device or regular file", cpath);
+		return (-1);
+	}
+	fnvlist_add_string(item, "path", cpath);
+	uint64_t wd = zfs_dev_is_whole_disk(rpath);
+	fnvlist_add_uint64(item, "whole_disk", wd);
+	return (0);
+}
+
+static
+int add_draid_data(nvlist_t *nvl, PyObject *item, uint64_t size, const char *ctype)
+{
+	PyObject *nsd, *nssp;
+	PyObject *nd, *nsp;
+	nsd = PyUnicode_FromString("draid_data_disks");
+	if (!nsd) {
+		PyErr_SetString(PyExc_RuntimeError,
+		    "Failed to create Python String");
+		return (-1);
+	}
+	nssp = PyUnicode_FromString("draid_spare_disks");
+	if (!nssp) {
+		Py_DECREF(nsd);
+		PyErr_SetString(PyExc_RuntimeError,
+		    "Failed to create Python String");
+		return (-1);
+	}
+	nd = PyDict_GetItem(item, nsd);
+	if (!PyLong_Check(nd)) {
+		PyErr_SetString(PyExc_TypeError,
+		    "Expected a list for key \'draid_data_disks\'");
+		Py_DECREF(nsd);
+		Py_DECREF(nssp);
+		return (-1);
+	}
+	nsp = PyDict_GetItem(item, nssp);
+	if (!PyLong_Check(nsp)) {
+		PyErr_SetString(PyExc_TypeError,
+		    "Expected a list for key \'draid_spare_disks\'");
+		Py_DECREF(nsd);
+		Py_DECREF(nssp);
+		return (-1);
+	}
+	uint64_t p = ctype[strlen(ctype) - 1] - '0';
+	uint64_t d = PyLong_AsUnsignedLong(nd);
+	if (PyErr_Occurred()) {
+		PyErr_SetString(PyExc_TypeError,
+			"Failed to convert to unsigned long");
+		Py_DECREF(nsd);
+		Py_DECREF(nssp);
+		return (-1);
+	}
+	uint64_t s = PyLong_AsUnsignedLong(nsp);
+	if (PyErr_Occurred()) {
+		PyErr_SetString(PyExc_TypeError,
+			"Failed to convert to unsigned long");
+		Py_DECREF(nsd);
+		Py_DECREF(nssp);
+		return (-1);
+	}
+	uint64_t g = 1;
+	while (g * (d + p) % (size - s) != 0)
+		g++;
+	fnvlist_add_string(nvl, "type", "draid");
+	fnvlist_add_uint64(nvl, "nparity", p);
+	fnvlist_add_uint64(nvl, "draid_ndata", d);
+	fnvlist_add_uint64(nvl, "draid_nspares", s);
+	fnvlist_add_uint64(nvl, "ngroups", g);
+	Py_DECREF(nsd);
+	Py_DECREF(nssp);
+	return (0);
+}
+
+static
+void free_nvlist_array(nvlist_t **nvlarr, uint64_t len)
+{
+	for (uint64_t i = 0; i < len; ++i)
+		fnvlist_free(nvlarr[i]);
+	free(nvlarr);
+}
+
+nvlist_t *make_vdev_tree(PyObject *topology, PyObject *props)
+{
+	uint64_t tlc = 0, nsp = 0, nl2c = 0;
+	char croot[128], ctype[128];
```

Comment:
Are there any constants we can use for these? Maybe `ZFS_MAXPROPLEN`?

Reply:
There isn't a constant for 128. Closest I have found and used is `ZAP_MAXNAMELEN` (256).